main/src/main/scala/sbt/Defaults.scala

```diff
@@ -330,9 +330,23 @@ object Defaults extends BuildCommon {
       val baseDir = baseDirectory.value
       val bases = unmanagedSourceDirectories.value
       val include = (includeFilter in unmanagedSources).value
-      val exclude = (excludeFilter in unmanagedSources).value
+      val exclude = (excludeFilter in unmanagedSources).value match {
+        case e =>
+          (managedSources in ThisScope).value match {
+            case l if l.nonEmpty =>
+              e || new FileFilter {
+                private val files = l.toSet
+                override def accept(pathname: File): Boolean = files.contains(pathname)
+                override def toString = s"ManagedSourcesFilter($files)"
+              }
+            case _ => e
+          }
+      }
       val baseSources =
-        if (sourcesInBase.value) Seq(new Source(baseDir, include, exclude, recursive = false))
+        if (sourcesInBase.value)
+          Seq(new Source(baseDir, include, exclude || new SimpleFileFilter({ f =>
```

Comment:
I can remove this diff if this commit is brought in after #4096 (which makes the recursive source filter out files not in the base path).

Reply:
OK, let's see how things land.